### PR TITLE
Service call() API

### PR DESCRIPTION
**Pull request by mikepersonick**
This is just a draft, for API review.

### CODE REVIEW

File: gremlin-core/src/main/java/org/apache/tinkerpop/gremlin/process/traversal/dsl/graph/GraphTraversal.java [hunk 1]
```diff
@@ -1320,6 +1322,81 @@
         return this.asAdmin().addStep(new MathStep<>(this.asAdmin(), expression));
     }
 
+    /**
+     * Map a {@link Property} to its {@link Element}.
+     *
+     * @return the traversal with an appended {@link ElementStep}.
+     * @see <a href="http://tinkerpop.apache.org/docs/${project.version}/reference/#element-step" target="_blank">Reference Documentation - Element Step</a>
+     * @since 3.6.0-incubating
```
Comment (spmallette) [inline gremlin-core/src/main/java/org/apache/tinkerpop/gremlin/process/traversal/dsl/graph/GraphTraversal.java:1330]: version should be just "3.6.0"

File: gremlin-core/src/main/java/org/apache/tinkerpop/gremlin/process/traversal/dsl/graph/GraphTraversal.java [hunk 2]
```diff
@@ -1320,6 +1322,81 @@
...
+     */
+    default <E> GraphTraversal<S, E> call(final String service, final Traversal<?, Map<?,?>> childTraversal) {
+        this.asAdmin().getBytecode().addStep(Symbols.call, service, childTraversal);
+        final CallStep<S,E> step = null == childTraversal ? new CallStep(this.asAdmin(), false, service) :
+                new CallStep(this.asAdmin(), false, service, new LinkedHashMap(), childTraversal.asAdmin());
+        return this.asAdmin().addStep(step);
+    }
+
+    /**
+     * Perform the specified service call with both static and dynanmic parameters produced by the specified child
```
Comment (spmallette) [inline gremlin-core/src/main/java/org/apache/tinkerpop/gremlin/process/traversal/dsl/graph/GraphTraversal.java:1383]: nit: spelling "dynamic" - perhaps also good to add a sentence to say how they merge will take place in the event of conflict, i.e. what happens when the same key is found in both maps?

File: gremlin-test/features/map/Call.feature [hunk 3]
```diff
@@ -0,0 +1,250 @@
...
+# under the License.
+
+@StepClassMap @StepCall @TinkerGraphOnly
+Feature: Step - Call
+
+  Scenario: g_call
+    Given the empty graph
+    And the traversal of
+      """
+      g.call()
```
Comment (mschmidt00) [inline gremlin-test/features/map/Call.feature:25]: nice, love this self-describing functionality

File: tinkergraph-gremlin/src/main/java/org/apache/tinkerpop/gremlin/tinkergraph/structure/TinkerHelper.java [hunk 4]
```diff
@@ -211,4 +218,52 @@ else if (edgeLabels.length == 1)
...
+     */
+    public static <E extends Element> Iterator<Property> search(final TinkerGraph graph, final String regex,
+                                                                final Optional<ElementType> type) {
+
+        final Supplier<Iterator<Element>> vertices = () -> IteratorUtils.cast(graph.vertices());
+        final Supplier<Iterator<Element>> edges = () -> IteratorUtils.cast(graph.edges());
+        final Supplier<Iterator<Element>> vertexProperties =
+                () -> IteratorUtils.flatMap(vertices.get(), v -> IteratorUtils.cast(v.properties()));
+
+        Iterator it;
```
Comment (mschmidt00) [inline tinkergraph-gremlin/src/main/java/org/apache/tinkerpop/gremlin/tinkergraph/structure/TinkerHelper.java:234]: Looking at this raises an interesting question: how would an external service provider, who does not have access to the node and edge properties, implement something like this? Assuming maximized efficiency, in the general case we wouldn't want to "always" fetch properties for the input type, on the other hand, if one doesn't don't do, the information to execute the service call may in some cases not be present. 

One option would be to design the interface differently, e.g. expect a map from nodes to their properties as input. Another option could be to tackle the problem architecturally, by providing information about whether or not properties are to be fetched in the service metadata (including optionally to use built-in functionality & interfaces to handle these situations).

I don't have a good answer here as well, but maybe you have any thoughts?

File: tinkergraph-gremlin/src/main/java/org/apache/tinkerpop/gremlin/tinkergraph/structure/TinkerServiceRegistry.java [hunk 5]
```diff
@@ -0,0 +1,308 @@
...
+         */
+        String DEGREE_CENTRALITY = "tinker.degree.centrality";
+    }
+
+    private final TinkerGraph graph;
+
+    public TinkerServiceRegistry(final TinkerGraph graph) {
+        this.graph = graph;
+        registerService(new SearchService(graph));
+        registerService(new DegreeCentrality(graph));
```
[the remark this answers is not in the document]
Comment (mschmidt00) [reply tinkergraph-gremlin/src/main/java/org/apache/tinkerpop/gremlin/tinkergraph/structure/TinkerServiceRegistry.java:78]: +1 from my side on not including them by default -- having them pre-registered may complicate things for external services building upon TP; might even be nice to have support for service registration via config file (no must, just a thought)

File: gremlin-core/src/main/java/org/apache/tinkerpop/gremlin/process/traversal/step/map/CallStep.java [hunk 6]
```diff
@@ -0,0 +1,303 @@
...
+                if (service.isBarrier()) {
+                    /*
+                     * Barrier service - gather all upstream input and call.
+                     */
+                    final TraverserSet<S> traverserSet = (TraverserSet<S>) this.traversal.getTraverserSetSupplier().get();
+                    this.starts.forEachRemaining(traverserSet::add);
+                    this.iterator = this.flatMap(traverserSet);
+                } else {
+                    /*
+                     * Streaming service, mark the next start and call.
```
Comment (mschmidt00) [inline gremlin-core/src/main/java/org/apache/tinkerpop/gremlin/process/traversal/step/map/CallStep.java:159]: Just to confirm: it looks like batching support (and the resulting requirement of grouping parameters, etc.) is still missing?

File: gremlin-core/src/main/java/org/apache/tinkerpop/gremlin/structure/service/Service.java [hunk 7]
```diff
@@ -0,0 +1,212 @@
...
+    enum Type {
+        /**
+         * Start the traversal with no upstream input.
+         */
+        Start,
+
+        /**
+         * Mid-traversal with streaming input.
+         */
+        Streaming,
```
[the remark this answers is not in the document]
Comment (mikepersonick) [reply gremlin-core/src/main/java/org/apache/tinkerpop/gremlin/structure/service/Service.java:60]: Multi-type services.

**Pull request closed without merging**